serverless/email/webhook/main.go

```diff
@@ -0,0 +1,372 @@
+
+var (
+	// env vars
+	sesSource              = os.Getenv("SOURCE_EMAIL_ADDR")
+	namespaceArn           = os.Getenv("EMAIL_UUID_NAMESPACE")
+	authTokensArn          = os.Getenv("AUTH_TOKENS")
+	authSecretsArn         = os.Getenv("AUTH_SECRETS")
+	configSet              = os.Getenv("SES_CONFIG_SET")
+	dynamoRoleArn          = os.Getenv("DYNAMODB_ROLE_ARN")
+	dynamoEndpoint         = os.Getenv("DYNAMODB_ENDPOINT")
+	sesDomainSourceArn     = os.Getenv("SES_DOMAIN_SOURCE_ARN")
+	sesDomainFromArn       = os.Getenv("SES_DOMAIN_FROM_ARN")
+	sesDomainReturnPathArn = os.Getenv("SES_DOMAIN_RETURN_PATH_ARN")
+
+	// tables
+	idempotencyTable = aws.String("idempotency")
+	unsubscribeTable = aws.String("unsubscribe")
+
+	// clients
+	dynamoClient         *dynamodb.Client
+	sesClient            *ses.Client
+	secretsManagerClient *secretsmanager.Client
+
+	namespaceSecretOutput   *secretsmanager.GetSecretValueOutput
+	authTokenSecretOutput   *secretsmanager.GetSecretValueOutput
+	authSecretsSecretOutput *secretsmanager.GetSecretValueOutput
+
+	// header names
+	tsHeaderName  = os.Getenv("TS_HEADER_NAME")
+	sigHeaderName = os.Getenv("SIG_HEADER_NAME")
+	keyHeaderName = os.Getenv("KEY_HEADER_NAME")
+)
+
+// validateSignature - validate the signature provided from partner
+func validateSignature(request events.APIGatewayProxyRequest) error {
+	// signingString => {ts}{method}{path}{body}
+	signingString := fmt.Sprintf("%s%s%s%s",
+		request.Headers[tsHeaderName], request.HTTPMethod, request.Path, request.Body)
+	signatureBytes, err := hex.DecodeString(request.Headers[sigHeaderName])
+	if err != nil {
+		return fmt.Errorf("failed to decode signature from headers: %w", err)
+	}
+
+	var valid bool
+	// sha256 hmac(apiSecret, signingString)
+	for _, apiSecret := range strings.Split(*authSecretsSecretOutput.SecretString, ",") {
+		mac := hmac.New(sha256.New, []byte(apiSecret))
+		mac.Write([]byte(signingString))
+		expectedMAC := mac.Sum(nil)
+		if hmac.Equal(signatureBytes, expectedMAC) {
+			valid = true
+			break
+		}
+	}
+
+	if !valid {
+		// invalid signature
+		return errors.New("invalid signature for webhook")
+	}
+	return nil
+}
+
+// handler takes the api gateway request and sends a templated email
+func handler(ctx context.Context) func(request events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) {
+	logger := logging.Logger(ctx, "webhook.handler")
+	return func(request events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) {
+		var (
+			authenticated  bool
+			apiKey, authOK = request.Headers[keyHeaderName]
+			apiKeyHash     = sha256.Sum256([]byte(apiKey))
+		)
+
+		// no api key in request
+		if !authOK {
+			return events.APIGatewayProxyResponse{
+				StatusCode: http.StatusUnauthorized,
+				Body:       http.StatusText(http.StatusUnauthorized),
+			}, nil
+		}
+
+		// check auth token against our comma seperated list of valid auth tokens
+		for _, token := range strings.Split(*authTokenSecretOutput.SecretString, ",") {
```

Comment:
What's the need for keeping this as a list of tokens here?

Reply:
This is to allow the partner to rotate API Keys without a hard cutover.

Reply:
resolving this one as well